### PR TITLE
Acceptance tests (must requirements)

### PR DESCRIPTION
**Pull request by JWillegers**
- setup to create new tests
- tests for scenario-delta-value graph
- tests for exporting (and importing) with JSON

### CODE REVIEW

File: atest/resources/helpers/modelgenerator.py [hunk 1]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
+    @keyword(name='Get Length All Traces')  # type:ignore
+    def get_length_all_traces(self, trace_info: TraceInfo) -> int:
+        return len(trace_info.all_traces)
+
+    # ============= #
+    # == HELPERS == #
+    # ============= #
+
+    @staticmethod
+    def __convert_to_state_info(scenario_name: str, keyvaluestr: str | None) -> tuple[ScenarioInfo, StateInfo]:
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: Does this only handle a single property? E.g. it can't do
prop1: key1=value1
prop2: key2=value2

Maybe it's fine for our testing purposes, but I feel there might be edge-case tests where one property is updated but another isn't? Or ensuring ordering of properties doesn't matter? Things like that.

Fine as-is, just wanted to put that out there

Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: Additionally, the name should be more something like `__convert_to_info_tuple` or similar, as it create scenario AND state info, not just state info.

File: atest/resources/helpers/modelgenerator.py [hunk 2]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
+                    found = True
+                    break
+
+            if not found:
+                return False
+
+        return True
+
+    @keyword(name='Backtrack')  # type:ignore
+    def backtrack(self, trace_info: TraceInfo, steps: int) -> TraceInfo:
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py:149]: This implementation uses internal trace info methods, which is fine, but I'd prefer using the outer update_trace function, extracting the proper scenario/state from the internal lists, and setting length to current trace length - steps. E.g.
```python
    @keyword(name='Backtrack')  # type:ignore
    def backtrack(self, trace_info: TraceInfo, steps: int) -> TraceInfo:
        scenario, state = trace_info.current_trace[-steps - 1]
        trace_info.update_trace(scenario, state, len(trace_info.current_trace) - steps)
        return trace_info
```
(this passes the existing tests)

File: atest/resources/visualisation.resource [hunk 3]
```diff
@@ -7,68 +7,155 @@ Library           Collections
...
     Variable Should Exist    ${trace_info}
+    ${trace_info} =    The Algorithm Inserts    ${trace_info}    ${scenario_name}    ${state_str}
 
-current trace ${current_trace} has a tuple 'scenario ${scenario}, state ${state}'
+test suite ${suite} has a trace with ${n} steps
     [Documentation]    *model info*
-    ...     :IN: trace_info.current_trace
-    ...     :OUT: trace_info.current_trace.append((${scenario}, ${state}))
-    Variable Should Exist   ${trace_info}
+    ...    :IN: trace_info.current_trace==${n}
```
Comment (Egietje) [inline atest/resources/visualisation.resource]: This works? Is this some robot voodoo magic? Comparing a list against a number? Or is this some unintended behavior that doesn't actually work but still passes somehow? Shouldn't it be `len(trace_info.current_trace)==${n}`?

File: atest/resources/visualisation.resource [hunk 4]
```diff
@@ -7,68 +7,155 @@ Library           Collections
...
+    [Documentation]    *model info*
+    ...    :IN: graph.name=${name}
+    ...    :OUT: graph.name=${name}
+    Variable Should Exist    ${graph}
+
+    ${result} =     Graph Contains Vertex With Text        ${graph}      ${scenario}       ${text}
+    Run Keyword If      "${result}" != "None"
+    ...    Fail     Fail: Graph contains '${scenario}' with "${text}"
+    
+graph ${name} has an edge from vertex '${start_title}' to ${vertex}'${end_title}'
```
Comment (Egietje) [inline atest/resources/visualisation.resource]: Why the weirdness/inconsistency with `#{vertex}`? Why require the word vertex in some places but not all?

File: robotmbt/visualise/networkvisualiser.py [hunk 5]
```diff
@@ -38,6 +38,16 @@
 MINOR_FONT_SIZE: int = 8
 
 
+class RawNodeInfo:
```
Comment (Egietje) [inline robotmbt/visualise/networkvisualiser.py]: This is an exact copy of the Node class below, surely you can just use that class?

File: robotmbt/visualise/networkvisualiser.py [hunk 6]
```diff
@@ -574,6 +589,15 @@ def _add_node_to_sources(node: Node, final_trace: list[str], node_source: Column
     node_label_source.data['y'].append(-node.y)
     node_label_source.data['label'].append(node.label)
 
+    return RawNodeInfo(
```
Comment (Egietje) [inline robotmbt/visualise/networkvisualiser.py]: Is there even a need to instantiate a new object here? You copy all attributes from Node anyway apart from the node_id.  Maybe move the conversion to a hashable id to `_create_layout()` (specifically, line 229), and create a dict from id to Node after that. Gets rid of a lot of changes to this class that aren't needed.

File: atest/resources/helpers/modelgenerator.py [hunk 7]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
+            node: Node | None = network_vis.node_dict[id]
+            return node.y
+        except KeyError:
+            return None
+
+    @keyword(name='Graph Contains Vertices')  # type:ignore
+    def scen_graph_contains_vertices(self, graph: AbstractGraph, vertices_str: str) -> bool | str:
+        attr: dict[str, str] = nx.get_node_attributes(graph.networkx, "label")
+
+        vertices = vertices_str.split(" ")
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: What about vertices with a space in their name? Maybe split on `' '` instead, then only the first vertex matched has to start with `'` and the last vertex has to end with `'`.

File: atest/resources/helpers/modelgenerator.py [hunk 8]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
+        if id is None:
+            return None
+
+        try:
+            node: Node | None = network_vis.node_dict[id]
+            return node.y
+        except KeyError:
+            return None
+
+    @keyword(name='Graph Contains Vertices')  # type:ignore
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: Since you're just checking the start of the label, maybe call this `Graph Contains Vertices Starting With`

Comment (Egietje) [reply atest/resources/helpers/modelgenerator.py]: For example, now it'll return true when checking for v1 even if all it contained was v11

File: atest/resources/helpers/modelgenerator.py [hunk 9]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
+    def backtrack(self, trace_info: TraceInfo, steps: int) -> TraceInfo:
+        trace_info.pushed = True
+        trace_info._pop(steps)
+        return trace_info
+
+    @keyword(name='Get Length Current Trace')  # type:ignore
+    def get_length_current_trace(self, trace_info: TraceInfo) -> int:
+        return len(trace_info.current_trace)
+
+    @keyword(name='Get Length All Traces')  # type:ignore
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: Maybe make this more explicit to what it represents, e.g. `Get Number of Backtracked Traces`

File: atest/resources/helpers/modelgenerator.py [hunk 10]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
+            prev_value: str = " ".join(splits[:-1]) if len(splits) > 1 else splits[0]
+            new_key: str = splits[-1]
+
+            keyvalues.append((prev_key, prev_value))
+            prev_key = new_key
+
+        return (ScenarioInfo(scenario_name), StateInfo._create_state_with_prop(domain, keyvalues))
+
+    @staticmethod
+    def __split_top_level(text):
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py:204]: This method hurts my soul but if it works it works

File: atest/resources/helpers/modelgenerator.py [hunk 11]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
+
+    @keyword(name="Vertices Are Connected")  # type:ignore
+    def vertices_connected(self, graph: AbstractGraph, node_key1: str | None, node_key2: str | None) -> bool:
+        if node_key1 is None or node_key2 is None:
+            return False
+        return graph.networkx.has_edge(node_key1, node_key2)
+
+    @keyword(name="Get Vertex Y Position")  # type:ignore
+    def get_y(self, graph: AbstractGraph, network_vis: NetworkVisualiser, node_title: str) -> int | None:
+        id: str | None = self.graph_contains_vertex_with_text(graph, node_title, text=None)
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: Maybe separate this logic out on a higher level? Passing in a node id instead of a node title, and requiring the resource file to separately obtain the ID? Makes for more proper error messages (I think) and reduces connected-ness of this internal logic

File: atest/resources/helpers/modelgenerator.py [hunk 12]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
-        return succes if repr(t1) == repr(t2) else fail
+    def compare_trace_info(self, t1: TraceInfo, t2: TraceInfo) -> bool:
+        return repr(t1) == repr(t2)
 
     @keyword(name='Delete File')  # type:ignore
-    def delete_json_file(self, filepath: str):
+    def delete_file(self, filepath: str):
         os.remove(filepath)
+
+    @keyword(name='Graph Contains No Text')  # type:ignore
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: I think this name is quite confusing if this is just an existence check. Also, why not just return a boolean?

File: atest/resources/helpers/modelgenerator.py [hunk 13]
```diff
@@ -30,54 +31,234 @@ def all_traces_contains_list(self, trace_info: TraceInfo) -> TraceInfo:
...
-        '''
-        succes = 'imported model equals exported model'
-        fail = 'imported models differs from exported model'
-        return succes if repr(t1) == repr(t2) else fail
+    def compare_trace_info(self, t1: TraceInfo, t2: TraceInfo) -> bool:
+        return repr(t1) == repr(t2)
 
     @keyword(name='Delete File')  # type:ignore
-    def delete_json_file(self, filepath: str):
+    def delete_file(self, filepath: str):
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py:75]: In before `Delete File system32`

File: atest/resources/visualisation.resource [hunk 14]
```diff
@@ -7,68 +7,155 @@ Library           Collections
 *** Keywords ***
 test suite ${suite}
```
Comment (Egietje) [inline atest/resources/visualisation.resource]: Why the duplicate functionality? This does the same as `test suite ${suite} has trace info ${trace}`.

File: atest/resources/visualisation.resource (hunk 3 above)
Comment (Egietje) [reply atest/resources/visualisation.resource]: Wait I just realised this doesn't access the actual trace_info, disregard this comment

File: atest/resources/visualisation.resource [hunk 15]
```diff
@@ -7,68 +7,155 @@ Library           Collections
...
+
+the algorithm inserts '${scenario_name}' with state "${state_str}"
+    [Documentation]    *model info*
+    ...    :IN: trace_info
+    ...    :OUT: trace_info.current_trace+=1
     Variable Should Exist    ${trace_info}
+    ${trace_info} =    The Algorithm Inserts    ${trace_info}    ${scenario_name}    ${state_str}
 
-current trace ${current_trace} has a tuple 'scenario ${scenario}, state ${state}'
+test suite ${suite} has a trace with ${n} steps
```
Comment (Egietje) [inline atest/resources/visualisation.resource]: Again a duplicate function, the same as `test suite ${suite} has ${n} steps in its current trace`.

File: atest/resources/visualisation.resource [hunk 16]
```diff
@@ -7,68 +7,155 @@ Library           Collections
...
+vertex '${start_vertex}' is placed above '${end_vertex}'
+    [Documentation]    *model_info*
+    ...    :IN: None
+    ...    :OUT: None
+    Variable Should Exist    ${graph}
+    Variable Should Exist    ${network_visualiser}
+    ${result1} =    Get Vertex Y position   ${graph}    ${network_visualiser}    ${start_vertex}
+    ${result2} =    Get Vertex Y position   ${graph}    ${network_visualiser}    ${end_vertex}
+    
+    # higher y is lower.... yeah I know. Confusing.
```
Comment (Egietje) [inline atest/resources/visualisation.resource]: Internally, the y gets inverted when passing it to Bokeh

File: atest/robotMBT tests/10__visualisation_representations/1__scenario-delta-value.robot [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+Vertex Scenario-Delta-Value graph
+    Given test suite s has a trace with 2 steps
+    When scenario-delta-value graph g is generated
+    Then graph g contains vertex 'start'
+    And graph g contains vertex 'A1' with text "attr: states = ['a1'], special='!'"
+    And graph g contains vertex 'A2' with text "attr: states = ['a1', 'a2']"
+    And graph g does not contain vertex 'A2' with text "attr: states = ['a1', 'a2'], special='!'"
+
+Edge Scenario-Delta-Value graph
+    Given test suite s has a trace with 2 steps
```
Comment (Egietje) [inline atest/robotMBT tests/10__visualisation_representations/1__scenario-delta-value.robot]: Technically this given doesn't represent that the thens should hold, it could be any trace with two steps, not necessarily with steps called A1 and A2 in that order

**Review by Egietje — CHANGES_REQUESTED** [review]

File: atest/resources/visualisation.resource (hunk 3 above)
Comment (JWillegers) [reply atest/resources/visualisation.resource]: Indeed, this doesn't access the current trace but is just a model state

File: atest/resources/visualisation.resource (hunk 14 above)
Comment (JWillegers) [reply atest/resources/visualisation.resource]: For most test suites this is used, as trace info is unneeded information. For export to JSON we want to explicitly add the trace info as it is used at a later point in the same suite

File: atest/resources/visualisation.resource (hunk 15 above)
Comment (JWillegers) [reply atest/resources/visualisation.resource]: Another case of slightly different wording to be more explicit with export to JSON test suite

Comment (JWillegers) [reply atest/resources/visualisation.resource]: They have been merged

File: atest/resources/visualisation.resource (hunk 14 above)
Comment (JWillegers) [reply atest/resources/visualisation.resource]: They have been merged

File: atest/resources/helpers/modelgenerator.py [hunk 18]
```diff
@@ -125,21 +125,18 @@ def get_y(self, graph: AbstractGraph, network_vis: NetworkVisualiser, node_title
...
-        vertices = vertices_str.split(" ")
-        for i in range(len(vertices)):
-            v = vertices[i]
-            if not v.startswith("'") or not v.endswith("'"):
-                return "vertices string was not properly formatted. Format: "'v1' 'v2' 'v3' 'v4'""
-            vertices[i] = vertices[i][1:-1]  # strip away quotes
-
-        for vname in vertices:
+        vertices = vertices_str.split("'")
+        for i in range(len(vertices), 2):
```
Comment (Egietje) [inline atest/resources/helpers/modelgenerator.py]: This is a bug, you've defined the start and stop of the range, not the stop and step:
`range(start, stop, step)`

Instead, do:
`range(0, len(vertices), 2)`

File: atest/resources/visualisation.resource
Comment (Egietje) [inline atest/resources/visualisation.resource]: Still has optional vertex

**Review by Egietje — CHANGES_REQUESTED** [review]

**Review by Egietje — APPROVED** [review]
Nice :)

File: atest/resources/helpers/modelgenerator.py (hunk 1 above)
Comment (Egietje) [reply atest/resources/helpers/modelgenerator.py]: Can be left out
